### PR TITLE
Multi GPU mode for codeparrot evaluation

### PR DESCRIPTION
**Pull request by liyongsea**
# What does this PR do?
This PR provdie a script to run codeparrot human eval in multi GPU mode

# Performance data
The performance is check using the following command
```
accelerate launch scripts/human_eval_mgpu.py --model_ckpt lvwerra/codeparrot-small --do_sample True --temperature 0.2 --top_p 0.95 --n_samples=40 --HF_ALLOW_CODE_EVAL="1" --device_int=0 --num_tasks=8 --batch_size=10
```
One should obtain
```
Results: {'pass@1': 0.1375, 'pass@10': 0.21251641317430792}
```

# Time benchmark
TODO

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: examples/research_projects/codeparrot/scripts/human_eval_mgpu.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+class TokenizeDataset(IterableDataset):
+    def __init__(self, tokenizer, dataset, max_length, n_tasks=None):
+        self.tokenizer = tokenizer
+        self.dataset = dataset
+        self.max_length = max_length
+        self.n_tasks = len(dataset) if n_tasks is None else n_tasks
+
+    def __iter__(self):
+        for task in range(self.n_tasks):
+            # why we need eos token? do we need strip?
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/human_eval_mgpu.py]: In the pretraining it can happen that not the full file is in the context of the model and thus there can be function definitions outside the context. However, files are concatenated with an EOS token during training, so if there is an EOS token the model "knows" that there is nothing before the current context it can use. I thought that way we could minimize the errors when the models tries to use undefined objects or unimported libraries. 

Never actually tested the difference but it shouldn't hurt the performance in any case.

File: examples/research_projects/codeparrot/scripts/human_eval_mgpu.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+
+def complete_code(pipe, prompt, num_completions=1, **gen_kwargs):
+    """Complete prompt with text generation pipeline and return num_completions."""
+    prompt = pipe.tokenizer.eos_token + prompt
+    code_gens = pipe(prompt, num_return_sequences=num_completions, **gen_kwargs)
+    return [first_block(code_gen["generated_text"][len(prompt) :]) for code_gen in code_gens]
+
+
+def complete_code_multi_gpu(model, tokenizer, dataloader, batch_size=20, **gen_kwargs):
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/human_eval_mgpu.py]: maybe we could call the function just `complete_code` as accelerate would also work on a single GPU/TPU or even CPU.

File: examples/research_projects/codeparrot/scripts/human_eval_mgpu.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    try:
+        _ = code_eval_metric.compute(references=[""], predictions=[[""]])
+    except ValueError as exception:
+        print(
+            'Code evaluation not enabled. Read the warning below carefully and then use `--HF_ALLOW_CODE_EVAL="1"` flag to enable code evaluation.'
+        )
+        raise exception
+
+    generations = complete_code_multi_gpu(model, tokenizer, human_eval_loader, batch_size=args.batch_size, **gen_kwargs)
+    for _ in range(args.n_samples // args.batch_size - 1):
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/human_eval_mgpu.py]: We could probably move that for loop inside the `complete_code` function, no? That way we could also avoid calling `complete_code` on two lines.

File: examples/research_projects/codeparrot/scripts/human_eval_mgpu.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        _gens = complete_code_multi_gpu(model, tokenizer, human_eval_loader, batch_size=args.batch_size, **gen_kwargs) 
+        for i, g in enumerate(_gens):
+            generations[i].extend(g)
+    references = []
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/human_eval_mgpu.py]: Fro your email I think you already figured out that you can wrap everything below here in a 
```python
if accelerator.is_main_process:
     ...
```

**Review by lvwerra — COMMENTED** [review]
Hi Jia, you are making great progress! I left a few comments. Also, when you make the PR to `huggingface/transformers` feel free to replace the existing file. The elegance of `accelerate` means that we can run the same code on a single or multiple GPUs. 

As a final test I think it would be good to reproduce the CodeParrot numbers on at least a 1-GPU and a 2-GPU setup so we know it is working as expected on a single and multiple GPUs. If you want I can also run those experiments on our infrastructure to save some money :)

File: examples/research_projects/codeparrot/scripts/human_eval_mgpu.py (hunk 1 above)
Comment (liyongsea) [reply examples/research_projects/codeparrot/scripts/human_eval_mgpu.py]: Very tricky .... indeed the model work in a wired way with strip ...
and the eos token is tricky as well

File: examples/research_projects/codeparrot/scripts/human_eval_mgpu.py (hunk 2 above)
Comment (liyongsea) [reply examples/research_projects/codeparrot/scripts/human_eval_mgpu.py]: agree !

File: examples/research_projects/codeparrot/scripts/human_eval_mgpu.py (hunk 3 above)
Comment (liyongsea) [reply examples/research_projects/codeparrot/scripts/human_eval_mgpu.py]: agree !

File: examples/research_projects/codeparrot/scripts/human_eval_mgpu.py (hunk 4 above)
Comment (liyongsea) [reply examples/research_projects/codeparrot/scripts/human_eval_mgpu.py]: Thank !

**Comment by liyongsea** [comment]
In the initial PR, TokenizeDataset iterate on the list of prompts [p1, p2, p3, p4, ....]; then p1, p2 are sent to different processes to leverage the muti GPU capacity. However, accelerate require the batches sent to multi GPU to have the same size. This is initially  handled by a padding of the EoS. Unfortunately, the codeparrot model has very different behavior according to the padding policy (@lvwerra is there a better way to pad the data?).

To avoid padding, in the next commit,  TokenizeDataset will send [p1, p1, ..., p1, p1, p2, p2, ...] where each prompt is sent n_copies times, where n_copies = n_samples // batch_size should be a multiple of the number of processes used in the evaluation

**Comment by liyongsea** [comment]
I discovered an error when launching the large model
```
ERROR:torch.distributed.elastic.multiprocessing.api:failed (exitcode: -9) local_rank: 3 (pid: 96970) of binary: /usr/local/bin/python
ERROR:torch.distributed.elastic.agent.server.local_elastic_agent:[default] Worker group failed
```
~@lvwerra any idea?~ -> no enough RAM when using accelerate 4 processes

**Comment by lvwerra** [comment]
> Unfortunately, the codeparrot model has very different behavior according to the padding policy (@lvwerra is there a better way to pad the data?)

Can we remove the padding before the generation? At this point the data should already be dispatched to the GPU and padding is no longer necessary. Would that solve your issue?

> no enough RAM when using accelerate 4 processes

Hmm, is this CPU or GPU memory that is running out? Maybe a smaller batch size would help? What kind of machine are you running this on? If it is consuming an absurd amount of CPU RAM maybe there is an issue in `Accelerate`.

File: examples/research_projects/codeparrot/scripts/human_eval_mgpu.py (hunk 1 above)
Comment (lvwerra) [reply examples/research_projects/codeparrot/scripts/human_eval_mgpu.py]: Oh actually didn't see the question about strip: The issue is that the tokenizer has a token `\n\t` and the model learns that after a function signature this token comes. However, in the HumanEval dataset by default there is just a `\n` and although the model also has a `\t` token it has never seen a `\n` + `\t` during pretraining since that is always replaced by the single `\n\t` token. Therefore this would be a very uncommon sequence and results in a weird behaviour in the model.

**Comment by liyongsea** [comment]
> Hmm, is this CPU or GPU memory that is running out? Maybe a smaller batch size would help? What kind of machine are you running this on? If it is consuming an absurd amount of CPU RAM maybe there is an issue in `Accelerate`.

I am talking about the CPU RAM. I will double check what was happening. With the large model, I manage to run it with 2 processes but not 4. I thought the CPU RAM might be 4 times high if I put 4 processes. Tell me if you have any intuitions, otherwise I will dig further

**Comment by lvwerra** [comment]
How much CPU and GPU memory do you have on your machine? It is quite surprising that you run out of CPU memory. Maybe we need to do some profiling to find the source. Or scale up the CPU memory on the machine.
